gadget/gadget.go

```diff
@@ -0,0 +1,630 @@
+	Target string `yaml:"target"`
+
+	// Image names the image, relative to gadget base directory, to be used
+	// for a 'bare' type structure
+	Image string `yaml:"image"`
+	// Offset the image is written at
+	Offset Size `yaml:"offset"`
+	// OffsetWrite describes a 32-bit address, within the volume, at which
+	// the offset of current image will be written. The position may be
+	// specified as a byte offset relative to the start of a named structure
+	OffsetWrite RelativeOffset `yaml:"offset-write"`
+	// Size of the image, when empty size is calculated by looking at the
+	// image
+	Size Size `yaml:"size"`
+
+	Unpack bool `yaml:"unpack"`
+}
+
+type VolumeUpdate struct {
+	Edition  editionNumber `yaml:"edition"`
+	Preserve []string      `yaml:"preserve"`
+}
+
+// GadgetConnect describes an interface connection requested by the gadget
+// between seeded snaps. The syntax is of a mapping like:
+//
+//  plug: (<plug-snap-id>|system):plug
+//  [slot: (<slot-snap-id>|system):slot]
+//
+// "system" indicates a system plug or slot.
+// Fully omitting the slot part indicates a system slot with the same name
+// as the plug.
+type Connection struct {
+	Plug ConnectionPlug `yaml:"plug"`
+	Slot ConnectionSlot `yaml:"slot"`
+}
+
+type ConnectionPlug struct {
+	SnapID string
+	Plug   string
+}
+
+func (gcplug *ConnectionPlug) Empty() bool {
+	return gcplug.SnapID == "" && gcplug.Plug == ""
+}
+
+func (gcplug *ConnectionPlug) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	snapID, name, err := parseSnapIDColonName(s)
+	if err != nil {
+		return fmt.Errorf("in gadget connection plug: %v", err)
+	}
+	gcplug.SnapID = snapID
+	gcplug.Plug = name
+	return nil
+}
+
+type ConnectionSlot struct {
+	SnapID string
+	Slot   string
+}
+
+func (gcslot *ConnectionSlot) Empty() bool {
+	return gcslot.SnapID == "" && gcslot.Slot == ""
+}
+
+func (gcslot *ConnectionSlot) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var s string
+	if err := unmarshal(&s); err != nil {
+		return err
+	}
+	snapID, name, err := parseSnapIDColonName(s)
+	if err != nil {
+		return fmt.Errorf("in gadget connection slot: %v", err)
+	}
+	gcslot.SnapID = snapID
+	gcslot.Slot = name
+	return nil
+}
+
+func parseSnapIDColonName(s string) (snapID, name string, err error) {
+	parts := strings.Split(s, ":")
+	if len(parts) == 2 {
+		snapID = parts[0]
+		name = parts[1]
+	}
+	if snapID == "" || name == "" {
+		return "", "", fmt.Errorf(`expected "(<snap-id>|system):name" not %q`, s)
+	}
+	return snapID, name, nil
+}
+
+func systemOrSnapID(s string) bool {
+	if s != "system" && len(s) != validSnapIDLength {
+		return false
+	}
+	return true
+}
+
+// ReadGadgetInfo reads the gadget specific metadata from gadget.yaml
+// in the snap. classic set to true means classic rules apply,
+// i.e. content/presence of gadget.yaml is fully optional.
+func ReadGadgetInfo(gadgetRootDir string, classic bool) (*Info, error) {
```

Comment:
I wonder if we can rename this now? `gadget.ReadInfo()` or just `gadget.Read()` or something?

Reply:
Ah, `ReadInfo()` sounds about right.